Lifecycle hook error catching

petebacondarwin opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
fix

**What is the current behavior? (You can also link to an open issue here)**

Errors in hooks are not dealt with very nicely
See #14444

**What is the new behavior (if this is a feature change)?**

Errors are caught and passed to the exceptionHandler in a reasonable way, without causing the application as a whole to crash

**Does this PR introduce a breaking change?**

No

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- ~~[ ] Docs have been added / updated (for bug fixes / features)~~

**Other information**:

petebacondarwin committed:
fix($compile): cope with `$onInit` hooks throwing

Previously, if one of these hooks threw an error, then the compilation
was terminated unceremoniously.

Closes #14444
Closes #14463

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -4032,6 +4072,92 @@ describe('$compile', function() {
+          $rootScope.$apply('a = 42');
+
+          // The first component's error should be logged
+          expect($exceptionHandler.errors[0].toString()).toEqual('Error: bad hook');
```
How come you didn't pop it ?

petebacondarwin replied:
a slip

gkalpak commented:
LGTM

petebacondarwin committed:
fix($compile): cope with `$onChanges` hooks throwing

Previously, if one of these hooks threw an error, then any other `$onChanges`
hooks that had been scheduled were not executed, nor was the queue cleaned up
properly.

Closes #14444
Closes #14463

petebacondarwin committed:
fix($compile): cope with `$onChanges` hooks throwing

Previously, if one of these hooks threw an error, then any other `$onChanges`
hooks that had been scheduled were not executed, nor was the queue cleaned up
properly.

Closes #14444
Closes #14463
